### ISSUE TITLE
"CredentialRequiresARNError" when using aws profiles.

### ISSUE THREAD

**Issue by DMonCode** [hashicorp/terraform-provider-aws#19897]
### Background:
I have been dealing with auth issue of aws provider >=3.0.0 for a while and have my teams using < 3.0.0 in our code
I have figured out part of the issue but wanted to update the list to make sure it is record as part of the issue was my lack of understanding changes that have been made verse the observed behavior.

TL:DR Version
* Initial release of AWS-SDK-GO did not support SSO. We maintainted workaround by using aws provider 2.70.0
* AWS-SDK-GO fixed/add SSO support. Have been trying to figure out how to make it work as aws provider 2.70.0 worked with our current config but 3.0.0+ did not.

My Findings:

* AWS provider >= 3.0.0 works closer to how the CLI works and the reference to "profile" will check `~/.aws/config` where as previous version only seems to look at `~/.aws/credentials.`
* Our use of `aws-adfs` sets up a selfreference problem in `~/.aws/config`. This was ultimately the cause of the error you will see below. That along with me not understanding all the changes.

### Terraform CLI and Terraform AWS Provider Version
Terraform v1.0.0
on darwin_amd64
+ provider registry.terraform.io/hashicorp/aws v3.45.0

NOTE: This issue has plagued me since 0.12.29 and provider hashicorp/aws v 3.0.0

### Affected Resource(s)

* `provider aws {}`
 

### Terraform Configuration Files

```hcl

variable auth {
  default  = {
    arn = "SPECIAL:ROLE:ARN"
    profile = "default"
    access_key = ""
    secret_key = ""
  }
}


provider aws {
  region = var.aws_region
  assume_role {
    role_arn = var.auth.arn
  }

  # There are two auth methods. Place values in one.
  # key/secret will override profile.
  
  ## Auth method 1: key/secret
  access_key = var.auth.access
  secret_key = var.auth.secret
  ## Auth method 2: profile 
  profile = var.auth.profile
}
```

NOTE: AWS Provider version 3.0.0+ this behavior has changes.

Historical Note: 
We made our provider statements this way so that local system could test against a sandbox with user permissions while stage/prod would run from TFE workspace that used set access/secret that the user didn't have. Users have ADFS based credentials and keys.

### Panic Output
This error would occur with any aws provider >=3.0.0 but works with 

```
 Error: error configuring Terraform AWS Provider: Error creating AWS session: CredentialRequiresARNError: credential type source_profile requires role_arn, profile default
```

Previous versions would honor "profile" and look only in the `~/.aws/credentials` for the name profile. It would use the value defined in `assumed_role` along with credentials found in `~/.aws/credentials`.

From the looks of it aws version >= 3.0.0 now interperpates `profile` as a values defined in `~/.aws/config`

### Expected Behavior

Authentication to work as it did in previous version of AWS provider.

### Actual Behavior
Authentication fails.

### Steps to Reproduce

Need aws auth files with:
'''
# ~/.aws/credentials
[default]
access_key_id = ########
secret_access_key_id = ######
'''

'''
# ~/.aws/config
[default]
region = us-east-2
source_profile = default
adfs_config.role_arn = arn:aws:iam::0000000000:role/ADFS-Team
adfs_config.ssl_verification = True
adfs_config.adfs_host = adfs.example.com
adfs_config.session_duration = 14400
adfs_config.provider_id = urn:amazon:webservices
adfs_config.sspi = False
adfs_config.u2f_trigger_default = True

[profile my-aws-env]
region = us-east-2
source_profile = default
role_arn = SPECIAL:ROLE:ARN
'''

1. `terraform apply`

### Important Factoids

Prior to AWS version 3.0.0 Terraform would use the value in "profile" and lookup an entry in `~/.aws/credentials` and pair it with `assumed_role` ARN.
With >=3.0.0 and terraform provider uses `~/.aws/config`.

However:
* AWS-GO-SDK does not allow for nested references. In the case of the config I have **my-aws-env** will try to use `source_profile = default` from `~/.aws/config` instead of from `~/.aws/credentials`. This sees to create a loop or just fails. As best I could tell the newer version of AWS-SDK-GO is using a more strict code interpritation of what `~/.aws/credentials` and `~/.aws/config` should look like.
* Because I work with a SAML based authntication I use `aws-adfs` to set the keys. If I use `aws-adfs login --adfs-host=mylogin.example.com` then it will generate an entry for "default" in both `~/.aws/credentials` and `~/.aws/config`. The problem is this will start our circular reference problem.

### References
* My Previous ticket: https://github.com/hashicorp/terraform-provider-aws/issues/14516
* AWS Auth configs reference to self referencing profiles: https://github.com/hashicorp/terraform-provider-aws/issues/9999
* AWS SDK GO lack of support for SSO: https://github.com/aws/aws-sdk-go-v2/issues/705
* AWS SDK GO issues with SSO: https://github.com/aws/aws-sdk-go-v2/issues/1222

**Comment by DMonCode** [comment]
### Solution:
* Name space your entries in `~/.aws/credentials` vs `~/.aws/config`
* Do not reference `source_profile` that can match an entry in `~/.aws/config`.
* All references for `source_profile` should only match entries in `~/.aws/credentials`

'''
# ~/.aws/credentials
[default]
access_key_id = ########
secret_access_key_id = ######
'''

'''
# ~/.aws/config
[profile adfsdefault]
region = us-east-2
source_profile = default
adfs_config.role_arn = arn:aws:iam::0000000000:role/ADFS-Team
adfs_config.ssl_verification = True
adfs_config.adfs_host = adfs.example.com
adfs_config.session_duration = 14400
adfs_config.provider_id = urn:amazon:webservices
adfs_config.sspi = False
adfs_config.u2f_trigger_default = True

[profile my-aws-env]
region = us-east-2
source_profile = default
role_arn = SPECIAL:ROLE:ARN
adfs_config.role_arn = arn:aws:iam::0000000000:role/ADFS-Team
adfs_config.ssl_verification = True
adfs_config.adfs_host = adfs.example.com
adfs_config.session_duration = 14400
adfs_config.provider_id = urn:amazon:webservices
adfs_config.sspi = False
adfs_config.u2f_trigger_default = True
'''


```hcl

variable auth {
  default  = {
    arn = "SOME:ARN:TO:ROLE"
    profile = "my-aws-env"
    access_key = ""
    secret_key = ""
  }
}


provider aws {
  region = var.aws_region
  assume_role {
    role_arn = var.auth.arn
  }

  # There are two auth methods. Place values in one.
  # key/secret will override profile.
  
  ## Auth method 1: key/secret
  access_key = var.auth.access
  secret_key = var.auth.secret
  ## Auth method 2: profile 
  profile = var.auth.profile
}
```

NOTE: Having a value for "profile" seems to override assume_role as it looks like it will not be used if access_key and secret_key are empty.

**Comment by DMonCode** [comment]
Their maybe some conflict upstream as specifications indicate that you should be able to nest values:
https://docs.aws.amazon.com/cli/latest/topic/config-vars.html 

> source_profile - The AWS CLI profile that contains credentials / configuration the CLI should use for the initial assume-role call. This profile may be another profile configured to use assume-role, though if static credentials are present in the profile they will take precedence. This parameter cannot be provided alongside credential_source.

I interpret this to mean I should be able to do 
```
# ~/.aws/credentials
[default]
access_key_id = 000000
secret_access_key_id = 00000sa00a

# ~/.aws/config
[default]
source_profile = default
region = us-east-2
output = json

[profile foo]
source_profile = default
```

However behavior from auth in given exampled with aws provider ~>3.0 breaks.

**Comment by DMonCode** [comment]
This breaks for profile "foo"
```
# ~/.aws/credentials
[default]
access_key_id = 000000
secret_access_key_id = 00000sa00a

# ~/.aws/config
[default]
source_profile = default
region = us-east-2
output = json

[profile foo]
source_profile = default
```

However this works:
```
# ~/.aws/credentials
[default]
aws_access_key_id = 000000
aws_secret_access_key_id = 00000sa00a
aws_session_token = 0123456789
aws_security_token = 0123456789 

[adfs]
aws_access_key_id = 000001
aws_secret_access_key_id = 00000sa001
aws_session_token = 01234567890
aws_security_token = 01234567890 

# ~/.aws/config
[default]
source_profile = default
region = us-east-2
output = json
adfs_config.ssl_verification = True
adfs_config.role_arn = arn:aws:iam::000111222333:role/ADFS-Role
adfs_config.adfs_host = login.example.com
adfs_config.adfs_user = domain\username
adfs_config.session_duration = 3600
adfs_config.provider_id = urn:amazon:webservices
adfs_config.sspi = False
adfs_config.u2f_trigger_default = True

[profile adfs]
region = us-east-2
output = json
adfs_config.ssl_verification = True
adfs_config.role_arn = arn:aws:iam::000111222333:role/ADFS-Role
adfs_config.adfs_host = login.example.com
adfs_config.session_duration = 3600
adfs_config.provider_id = urn:amazon:webservices
adfs_config.sspi = False
adfs_config.u2f_trigger_default = True

[profile foo]
source_profile = adfs
role_arn = arn:aws:iam:333222111000:role/Admin_Role
output = json

```

NOTE: 
* `[default]` in `~/.aws/config` has a source_profile but `[profile adfs]`does not. Both are created running `aws-adfs login --adfs-host=login.example.com` however the "adfs" one is ran with the `--profile=adfs` flag.

I do not understand why this works other than "default" is slightly different but looks to be to spec.

**Comment by DMonCode** [comment]
Closing issue. Will clarify my finding in separate post.

TL;DR profile behavior's with TFE and adfs access keys seem to have some quarks.
